src/lib/SaferNames/Syntax.hs

```diff
@@ -107,43 +100,53 @@ data Expr n =
  | Hof (Hof n)
    deriving (Show)
 
+data AtomBinderInfo (n::S) =
+   LetBound LetAnn (Expr n)
+ | LamBound Arrow
+ | PiBound
+ | MiscBound
+ | InferenceName
+
+data TypedBinderInfo (n::S) = TypedBinderInfo (Type n) (AtomBinderInfo n)
+
 data Decl n l = Let LetAnn (Binder n l) (Expr n)
                 deriving (Show)
 
-data DataConRefBinding n l = DataConRefBinding (Binder n l) (Atom n)
-                             deriving Show
+type AtomName   = Name TypedBinderInfo
 
-type AltP (e::E) = Abs (Nest Binder) e :: E
-type Alt = AltP Block                  :: E
+data Binder (n::S) (l::S) =
+  (:>) (NameBinder TypedBinderInfo n l) (Type n)  deriving Show
```

Comment:
I'm a bit confused as to why do we put the type both in the binder and in the `TypedBinderInfo`. Is it just to mimic how we used to construct `Bindings` before this change?

Reply:
A `NameBinder TypedBinderInfo n l` doesn't actually carry a `TypedBinderInfo`. It just promises that the name it binds is a `Name TypedBinderInfo l`, which can be queried in a `Scope l` to give a `TypedBinderInfo l`. We store an actual `Type n` alongside it so that when we go under a binder, we can add the type to the scope to fulfill that promise.